### PR TITLE
Third module/first task

### PR DESCRIPTION
**Pull request by PrimoTibalt**

### CODE REVIEW

File: ModuleThreeFirstTaskConsole/FileEventArgs.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+{
+    /// <summary>
+    /// Arguments for founded file.
+    /// </summary>
+    public class FileSystemEventArgs : EventArgs
+    {
+        /// <summary>
+        /// Information about file or directory.
+        /// </summary>
+        public FileSystemInfo Info { get; set; }
```
Comment (alexander-trishin) [inline ModuleThreeFirstTaskConsole/FileEventArgs.cs]: This property should be read-only.

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    {
+        private readonly DirectoryInfo fileSystem;
+        private readonly Predicate<FileSystemInfo> filter;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="FileSystemVisitor"/> class.
+        /// </summary>
+        /// <param name="fileSystem">Files and directories.</param>
+        /// <param name="filter">Filter of files and directories.</param>
+        public FileSystemVisitor(DirectoryInfo fileSystem, Predicate<FileSystemInfo> filter)
```
Comment (alexander-trishin) [inline ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: Please, use `IFileSystem` from `System.IO.Abstractions` NuGet package instead of `DirectoryInfo`

Comment (alexander-trishin) [inline ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: `Predicate` delegate is deprecated, please use `Func<T, bool>` instead.

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        private string GetNameWithoutLongPath(FileSystemInfo file)
+        {
+            return file.FullName.Replace(fileSystem.FullName, string.Empty);
+        }
+
+        private void CheckInputOfConstructor(DirectoryInfo fileSystem)
+        {
+            if (fileSystem is null)
+            {
+                throw new NullReferenceException(nameof(fileSystem));
```
Comment (alexander-trishin) [inline ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: Should be `ArgumentNullException`

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+            }
+            else if (Directory.Exists(fileSystem.FullName) == false)
```
Comment (alexander-trishin) [inline ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: `else` keyword is redundant

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        }
+
+        /// <summary>
+        /// To get name with directories started from provided fileSystem.
+        /// </summary>
+        /// <param name="file">File.</param>
+        /// <returns>Correctly cut name.</returns>
+        private string GetNameWithoutLongPath(FileSystemInfo file)
+        {
+            return file.FullName.Replace(fileSystem.FullName, string.Empty);
```
Comment (alexander-trishin) [inline ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: Why not to use `file.Name + file.Extension`? 😄

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        }
+
+        /// <summary>
+        /// Returns files from current directory.
+        /// </summary>
+        /// <param name="directoryInfo">Current direcotry.</param>
+        /// <returns>Files from directory.</returns>
+        private async IAsyncEnumerable<FileInfo> GetFilesFromCurrentDirectory(DirectoryInfo directoryInfo)
+        {
+            var filesTask = await Task.Run(directoryInfo.EnumerateFiles);
```
Comment (alexander-trishin) [inline ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: using `async` in this case will only worsen performance

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        {
+            FilteredFileFound?.Invoke(this, args);
+            Stoped = args.Stop ? args.Stop : Stoped;
+        }
+
+        /// <summary>
+        /// Invoke methods from FilteredDirectoryFound event.
+        /// </summary>
+        /// <param name="args"></param>
+        protected virtual void OnFilteredDirectoryFound(FileSystemEventArgs args)
```
Comment (alexander-trishin) [inline ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: The same method as for `OnFilteredFileFound` and others with type `(FileSystemEventArgs args) => void`. All of them can be refactored into one.

**Review by alexander-trishin — CHANGES_REQUESTED** [review]
Hi Anton :)

👍 for WPF app, but there are a few issues to fix. Also I'm still waiting for unit tests.

File: ModuleThreeFirstTaskConsole/FileEventArgs.cs (hunk 1 above)
Comment (PrimoTibalt) [reply ModuleThreeFirstTaskConsole/FileEventArgs.cs]: Fixed in cea260a

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs (hunk 2 above)
Comment (PrimoTibalt) [reply ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: Fixed in cea260a

Comment (PrimoTibalt) [reply ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: Fixed in cea260a

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs (hunk 3 above)
Comment (PrimoTibalt) [reply ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: Fixed in cea260a

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs (hunk 4 above)
Comment (PrimoTibalt) [reply ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: Fixed in cea260a

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs (hunk 5 above)
Comment (PrimoTibalt) [reply ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: The result isn't the same. I would like to return the file's path from the initial path provided in the constructor.

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs (hunk 6 above)
Comment (PrimoTibalt) [reply ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: Fixed in cea260a

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs (hunk 7 above)
Comment (PrimoTibalt) [reply ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: Fixed in cea260a

File: ModuleThreeFirstTaskConsole/FileSystemEventArgs.cs [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+{
+    /// <summary>
+    /// Arguments for founded file.
+    /// </summary>
+    public class FileSystemEventArgs : EventArgs
+    {
+        /// <summary>
+        /// Information about file or directory.
+        /// </summary>
+        public IFileSystemInfo Info { get; set; }
```
Comment (alexander-trishin) [inline ModuleThreeFirstTaskConsole/FileSystemEventArgs.cs]: This property is still not read-only

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+using System.IO.Abstractions;
+
+namespace ModuleThreeFirstTaskConsole
+{
+    /// <summary>
+    /// Can read filesystem and show documents and packages in linear form.
+    /// </summary>
+    public class FileSystemVisitor
+    {
+        private const string STANDARDPATH = @"C:\";
```
Comment (alexander-trishin) [inline ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: This is true only for Windows OS.

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="FileSystemVisitor"/> class.
+        /// </summary>
+        /// <param name="fileSystem">Files and directories.</param>
+        /// <param name="filter">Filter of files and directories.</param>
+        /// <param name="initialPath">Place in filesystem where to start searching.</param>
+        public FileSystemVisitor(
+            IFileSystem fileSystem,
+            Func<IFileSystemInfo, bool> filter,
+            string initialPath)
```
Comment (alexander-trishin) [inline ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: This greatly restricts the use of the class. Why couldn't we add a parameter to the Search method?

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+            _filter = filter is null ? x => true : filter;
+            _initialPath = initialPath ?? STANDARDPATH;
+            _initialPath = _fileSystem.Directory.Exists(initialPath) ? initialPath : throw new DirectoryNotFoundException("You entered path that doesn't exist.");
+            Stoped = true;
+        }
+
+        /// <summary>
+        /// Events that happen at the begin of search.
+        /// </summary>
+        public event EventHandler<IFileSystemInfo> SearchStarted;
```
Comment (alexander-trishin) [inline ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: `EventHandler` is deprecated. Use `Action<T>` instead.

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+
+            OnSearchEnded(directory);
+        }
+
+        /// <summary>
+        /// Returns every single directory and filtered files from sprecified directory.
+        /// </summary>
+        /// <param name="directory">Provided directory.</param>
+        /// <returns>FileInfo of each file and directory.</returns>
+        private async IAsyncEnumerable<IFileSystemInfo> GetAllFromCurrentDirectory(IFileSystemInfo directory)
```
Comment (alexander-trishin) [inline ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: asynchronous stream is still useless here

**Review by alexander-trishin — COMMENTED** [review]
Hi Anton :) It looks better now, but I'm still waiting for the task about adding unit tests.

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs (hunk 12 above)
Comment (PrimoTibalt) [reply ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: Fixed in a883af4

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs (hunk 11 above)
Comment (PrimoTibalt) [reply ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: Fixed in a883af4

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs (hunk 10 above)
Comment (PrimoTibalt) [reply ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: Fixed in a883af4

File: ModuleThreeFirstTaskConsole/FileSystemVisitor.cs (hunk 9 above)
Comment (PrimoTibalt) [reply ModuleThreeFirstTaskConsole/FileSystemVisitor.cs]: Fixed in a883af4

File: ModuleThreeFirstTaskConsole/FileSystemEventArgs.cs (hunk 8 above)
Comment (PrimoTibalt) [reply ModuleThreeFirstTaskConsole/FileSystemEventArgs.cs]: Fixed in a883af4

**Review by alexander-trishin — APPROVED** [review]

**Pull request merged**